### ISSUE TITLE
`stack upgrade` overrides Stack executable file not named `stack` when called as `stack`

### ISSUE THREAD

**Issue by andreasabel** [commercialhaskell/stack#6630]
I think `stack upgrade` should not overwrite any _versioned_ stack binary (`stack-x.y.z`):
```
$ stack upgrade
Current Stack version: 2.15.7; available download version: 3.1.1.
Newer version detected, downloading.
Querying for archive location for platform: osx-x86_64-static.
Downloading from:
https://github.com/commercialhaskell/stack/releases/download/v3.1.1/stack-3.1.1-osx-x86_64.tar.gz.
Download complete, testing executable.
Version 3.1.1, Git revision 8127279fb48012945f47f73167a5ecbce5692965 x86_64 hpack-0.37.0
New Stack executable available at: /Users/abel/.local/bin/stack.
Also copying Stack executable to: /Users/abel/.ghcup/bin/stack-2.15.7.
Stack executable copied successfully!

$ /Users/abel/.ghcup/bin/stack-2.15.7 --version
Version 3.1.1, Git revision 8127279fb48012945f47f73167a5ecbce5692965 x86_64 hpack-0.37.0
```

**Comment by mpilgrem** [comment]
That is interesting. I've been assuming (wrongly, it seems) that it was the GHCup `stack` executable that gets overwritten, not the 'real' Stack executable (albeit, one provided provided by GHCup). That is because (from Stack's online documentation):

> ... if the current running Stack executable is named `stack` (or, on Windows, `stack.exe`) (this is case insensitive), an existing binary distribution will replace it.

If the documentation is correct (to check!) that means that the running executable is named `stack` but the running executable's file is named something like `stack-2.15.7`.

It seems to me reasonable that `--only-local-bin` should be the default if the running executable has a *file* name other than `stack`.

**Comment by andreasabel** [comment]
There is a symlink `stack -> stack-2.15.7`: 
```
$ l ~/.ghcup/bin/stack
lrwxr-xr-x  1 abel  staff  12 May 20 17:13 /Users/abel/.ghcup/bin/stack -> stack-2.15.7
```
I guess in this case `stack upgrade` should overwrite the symlink source (`stack`), not the symlink target (`stack-2.15.7`).

**Comment by mpilgrem** [comment]
In module `Stack` we use `System.Environment.getProgName` (which finds its way to `upgrade = addCommand''` in module `Stack.CLI`). and in `Stack.Setup.downloadStackExe` we use `System.Environment.getExecutablePath`.

I don't think that I had appreciated that the two could be inconsistent when I implemented https://github.com/commercialhaskell/stack/pull/6119.

That is, what you are requesting is what I was trying to implement. I'll seek to make a change that caters for the command to run Stack and the executable's file name being different.
